redbpf/src/lib.rs

```diff
@@ -485,10 +597,56 @@ impl UProbe {
                     perf::open_uretprobe_perf_event(&path, offset + sym_offset, pid)?
                 }
             };
-            perf::attach_perf_event(fd, pfd)
+            let ret = perf::attach_perf_event(fd, pfd);
+            if ret.is_ok() {
+                self.attachment_points.push(UProbeAttachmentPoint {
+                    fn_name: fn_name.and_then(|fn_name| Some(fn_name.to_owned())),
+                    offset,
+                    target: target.to_owned(),
+                    pid,
+                    pfd,
+                });
+            } else {
+                libc::close(pfd);
+            }
+            ret
         }
     }
 
+    /// Detach the `uprobe` or `uretprobe`
+    ///
+    /// This method is not needed to be called manually because all attachment
+    /// points are detached and closed automatically when `UProbe` is
+    /// dropped. But this method provides a feature for detaching a bpf program
+    /// from uprobe event selectively.
+    ///
+    /// # Example
+    /// ```no_run
+    /// use redbpf::Module;
+    /// let mut module = Module::parse(&std::fs::read("file.elf").unwrap()).unwrap();
+    /// let uprobe = module.uprobe_mut("count_strlen").expect("bpf program not found");
+    /// uprobe.attach_uprobe(Some(&uprobe.name()), 0, "/lib/x86_64-linux-gnu/libc-2.30.so", None).unwrap();
+    /// // do some stuff...
+    /// uprobe.detach_uprobe(Some(&uprobe.name()), 0, "/lib/x86_64-linux-gnu/libc-2.30.so", None);
+    /// ```
+    pub fn detach_uprobe(
+        &mut self,
+        fn_name: Option<&str>,
+        offset: u64,
+        target: &str,
+        pid: Option<pid_t>,
+    ) -> Result<()> {
+        // bpf program is detached from perf event and the perf event is closed
+        // by dropping UProbeAttachmentPoint
+        self.attachment_points.retain(|ap| {
+            ap.fn_name.as_ref().and_then(|x| Some(x.as_str())) != fn_name
```

Comment:
`and_then(|x| Some(...))` is probably nicer as a `map(...)`.

Reply:
Ok. `map` looks more concise.